BIP 54: Consensus Cleanup (#1800)
bitcoin/bips: bip-0054.md

```
  BIP: 54
  Layer: Consensus (soft fork)
  Title: Consensus Cleanup
  Author: Antoine Poinsot <mail@antoinep.com>
          Matt Corallo <bips@bluematt.me>
  Comments-Summary: No comments yet.
  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0054
  Status: Draft
  Type: Standards Track
  Created: 2024-04-11
  License: CC0-1.0
```

## Abstract

This document proposes new consensus rules in order to fix the timewarp attack, reduce the worst
case block validation time, prevent Merkle tree weaknesses, and avoid duplicate transactions without
[bip-0030][BIP30] validation.

## Motivation

This proposal addresses a number of long-standing vulnerabilities and weaknesses in the Bitcoin

> protocol. Bundling these fixes together allows to overcome the fixed cost of deploying a Bitcoin

Reviewer: The sentence "Bundling..." is irrelevant and the claim seems unsubstantiated to erroneous. How does bundling BIPs constitute a material difference in the deployment cost; the activations of both SegWit and Taproot consisted of multiple BIPs. Suggest dropping the sentence entirely.

Author: You are misinterpreting the sentence. The bundle refers to the proposal, not the BIP.

soft fork.

> The timewarp bug permits a majority hashrate attacker to arbitrarily increase the block rate,

Reviewer: The timewarp bug hasn't been referenced prior to here in this document.

Suggest beginning with a summary listing of terms/bugs and their definitions.

Author: Fair enough. Added a link to the description of the bug.

allowing them to steal block subsidy from future miners and increase validation costs to nodes that
have to deal with the increased average transaction rate. By strategically setting the block
timestamp, the timewarp bug lets miners bring down the difficulty to its minimum within 38 days of
starting the attack. The existence of this bug not only significantly empowers a 51% attacker, but
also makes it notably harder to reason about miners' incentives. It could indeed be in the interest
of short-sighted miners as well as short-sighted users to exploit this vulnerability in a small
enough proportion to increase the block rate without fatally hurting the network, as the effectively

> increased block space would — all other things equal — bring fee rates down for users.

Reviewer: ```suggestion
increased block space would — all other things being equal — bring fee rates down for users.
```

> increased block space would - all others things equal - bring fee rates down for users.

Author: Thanks, done.

> Specially crafted blocks may be expensive to process, with validation times ranging from a few

Reviewer: Suggest separating the different issues in the motivation section with individual headers to improve clarity. Better yet, a focused BIP per issue and its mitigation.

In computing a block's Merkle root, a 64-byte transaction can be interpreted as an intermediate
node in the tree in addition to a leaf. This makes it possible to fake inclusion proofs by
pretending a 64-byte block transaction is an inner node, as well as to pretend the inner nodes on
one level of the tree are the actual block transactions.

Since [bip-0034][BIP34] activation, explicit [bip-0030][BIP30] validation is not necessary until
block height 1,983,702[^0].  Mandating new coinbase transactions be different from the early
[bip-0034][BIP34] violations makes it possible to get rid of [bip-0030][BIP30] validation forever.
Besides its unnecessary cost, another downside of [bip-0030][BIP30] validation is that it cannot be
performed by Utreexo clients. Finally, leveraging the coinbase transaction's `nLockTime` field
allows applications to recover the block height corresponding to a coinbase transaction without
having to parse Script.

## Specification

For all blocks after activation the following new rules apply.

Given a block at height `N`:
- if `N % 2016` is equal to 0, the timestamp of the block must be set to a value higher than or

>   equal to the value of the timestamp of block at height `N-1` minus 7200;

Reviewer: Agree with @murchandamus.

- if `N % 2016` is equal to 2015, the timestamp of the block must be set to a value higher than
  or equal to the value of the timestamp of the block at height `N-2015`.

A limit is set on the number of potentially executed signature operations in validating a
transaction. It applies to all transactions in the block except the coinbase transaction[^1]. For
each input in the transaction, count the number of `CHECKSIG` and `CHECKMULTISIG` in the input
scriptSig and previous output's scriptPubKey, including the P2SH redeemScript. The accounting is the
same as for [bip-0016][BIP16 specs]: a `CHECKSIG`/`CHECKSIGVERIFY` operation accounts for 1 and a
`CHECKMULTISIG`/`CHECKMULTISIGVERIFY` accounts for the number of public keys associated, or 20 if
the number of public keys is greater than 16. A `CHECKMULTISIG`/`CHECKMULTISIGVERIFY` not directly
preceded by a minimally-pushed number between 1 and 16 (included) accounts for 20. If the
total is strictly higher than 2500, the transaction is invalid.

> Transactions whose witness-stripped serialized size is exactly 64 bytes are made invalid.

Author: Sure. Done.

proposal][BIP-XXXX] to fix timewarp used a ten-minute grace period instead, also adopted for
[testnet4][BIP94 timewarp]. Out of an abundance of caution and because it only trivially worsens the
block rate increase under attack, a two-hour grace period is used here[^4].

Disabling some Script operations and functionalities was [previously proposed][BIP-XXXX] to reduce
the worst case block validation time but was met with resistance due to confiscation concerns[^5]. A
delicate balance needs to be struck between minimizing the confiscation risks of a mitigation, even
if merely theoretical, and bounding the costs one could impose on all other users of the system. To
this effect a limit on the number of potentially executed signature operations pinpoints exactly the
harmful behaviour, leaving maximum flexibility in how Script functionalities may have been used.
Such a limit reduces the worst case block validation time by a factor of 40 and drastically
increases the preparation cost of an attack to make it uneconomical for a miner[^6]. The 2500
value was chosen as the tightest value that did not make any non-pathological standard transaction

> invalid[^7].

Author: Thanks, done.

In the presence of 64-byte transactions a block header's Merkle root may be valid for different
sets of transactions. This is because in the Merkle tree construction a 64-byte transaction may be
interpreted as the catenation of two 32-byte hashes, or the catenation of two 32-byte hashes may
be interpreted as a transaction. The former allows to fake a block inclusion proof and the latter
makes it such that for a valid block the Merkle root in the block header is not a unique identifier
for the corresponding list of valid transactions[^8]. 64-byte transactions can only contain a

> scriptPubKey that lets anyone spend the funds, or burns them.  They have also been non-standard for

Author: Done.

> 6 years at the time this is written. It was suggested that the known vulnerabilities could instead be

Author: Removed the relative date entirely and just stuck with "since 2019".

mitigated by committing to the Merkle tree depth in the header's version field[^9]. The authors
believe it is preferable to address the root cause by invalidating 64-byte transactions. This
approach also fixes the vulnerability without developers of SPV verifiers having to implement the
mitigation or to know it is necessary in the first place.

Several blocks prior to [bip-0034][BIP34] activation contain a coinbase transaction whose scriptSig
contains a valid [bip-0034][BIP34] commitment to a future block height. This offers an opportunity
to duplicate these coinbase transactions in the future[^10] and for this reason [bip-0030][BIP30]

> validation will need to be re-activated from block 1,983,702. A simple way to prevent this is to

Author: Sure, but if you go this route you could even say technically it only have to be reactivated if not all of the the corresponding coinbase transaction's outputs were spent along with a non-duplicable input.

mandate that future coinbase transactions vary from coinbase transactions before [bip-0034][BIP34]
activation. There are multiple ways of achieving this, but setting and enforcing the timelock for
the coinbase transaction makes it so all coinbase transactions past Consensus Cleanup activation
could not have been valid before this height and therefore cannot be a duplicate[^11].

## Backward compatibility

This proposal only tightens the block validation rules: there is no block that is valid under the
rules proposed in this BIP but not under the existing Bitcoin consensus rules. As a consequence

> these changes are backward-compatible with unupgraded node software. That said, the authors strongly

Reviewer: ```suggestion
these changes are backward-compatible with non-upgraded node software. That said, the authors strongly
```

Author: Done.

encourage node operators to upgrade in order to fully validate all consensus rules.

## Miner forward compatibility

> Bitcoin Core version [0.16.1][Core 0.16.1] and later will not relay nor create block templates which

Reviewer: Use either "neither...nor" or "not...or", and "that" instead of "which"

```suggestion
Bitcoin Core version [0.16.1][Core 0.16.1] and later will neither relay nor create block templates that
```

Author: Done.

> include 64-byte transactions.

Author: It was done on purpose because it seemed to be "nicer" by increasing Bitcoin Core version number. But now that you point it it does seem clearer to keep the same ordering as in the other sections. Changed in latest push.

> Bitcoin Core version [29.0][Core 29.0] and later will not generate a block template which violates

Reviewer: ```suggestion
Bitcoin Core version [29.0][Core 29.0] and later will not generate a block template that violates
```

Author: Done.

the timestamp restrictions introduced in this BIP. Although it would be extremely unlikely due to
the grace period used in this proposal, miners should use the `curtime` or `mintime` field from the
`getblocktemplate` result for their block's timestamp to make sure they always create blocks valid
according to this proposal. Note this is not a new requirement: using a timestamp lower than the
`mintime` field from the `getblocktemplate` result already leads to creating an invalid block.

> Bitcoin Core as of version 29.0 may relay and create a block template including a transaction which

Reviewer: ```suggestion
Bitcoin Core as of version 29.0 may relay and create a block template that includes a transaction that
```

Author: Done.

violates the signature operations limit introduced in this BIP. A newer version of Bitcoin Core

> which makes this type of transaction non-standard should be widely adopted before this soft fork is

Reviewer: ```suggestion
that makes this type of transaction non-standard should be widely adopted before this soft fork is
```

(Is this a SHOULD, or a MUST?)

Author: Done. "Should" is good here.

considered for activation.

The coinbase transaction is usually crafted by mining pool software. To the best of the authors'
knowledge, there does not exist an open source reference broadly in use today for such software.
We encourage mining pools to update their software to craft coinbase transactions that are
forward-compatible with the changes proposed in this BIP.

## Acknowledgements

> This document builds upon an [earlier proposal][BIP-XXXX] by one of the authors.

Reviewer: It seems odd not to name the author here.

```suggestion
This document builds upon an [earlier proposal][BIP-XXXX] by Matt Corallo.
```

Author: (Took your suggestion either way.)
